contribution-guidelines.md

```diff
@@ -0,0 +1,147 @@
+
+For Go development, we recommend using [go-swagger][6] to generate and serve interactive API documentation. Note that
+[go-swagger][6] can be used in two modes. One mode generates source code for the microservice from the [OpenAPI][2] API
+specification. This can provide a quick way to get started on an API, but in our experience, maintaining microservices
+generated in this manner can be cumbersome. We've found that the better option is to generate the [OpenAPI][2]
+definition from the source code. You can find documentation for this approach in the [go-swagger documentation][9]. A
+few Go microservices used by Terrain, including [notifications][7] and [de-stats][8] use this approach successfully. See
+the `Makefile` in each repository for examples of the commands used to generate the API specifications. One important
+caveat is that when you're using this approach, the repository must be checked out into your GOPATH. For example, if you
+wish to do work on the notifications service, the repository should be checked out into
+`$GOPATH/src/github.com/cyverse-de/notifications`.
+
+## Robustness
+
+Robustness refers to the fault tolerance of an endpoint. In general, DE service endpoints should respond to errors as
+cleanly as possible. Care and consideration should be given when determining what to to in the event of an error. For
+example, what happens when the endpoint can't connect to the database or times out while reading a response from another
+service? In many cases, the error can be reported to the caller, but that might not always be the case. It may also be
+the case that we can report the error to the caller, but some cleanup has to be done beforehand.
+
+Sometimes, the service itself cannot respond to an error condition. For example, if a host that a service is running on
+encounters a network failure. In these cases, the service should be restarted automatically as soon as possible. For
+cases where the host itself encounters an error, merely having the service running inside Kubernetes will help. For
+other failure conditions, we have to provide a way for Kubernetes to detect when something is wrong with the
+service. Having useful and reliable health and readiness checks for each service will help to ensure that services will
+be restarted if something goes wrong.
+
+## Debugability
+
+Debugability means that callers should be able to identify and respond to errors. For REST interfaces, this means that
+HTTP status codes should be correct and that error response bodies should contain enough information for callers to
+identify and correct the problem. For example, if a resource isn't found, an HTTP endpoint should return status
+code 404. If there's a problem with the request then the status code should be in the 400 series, and a response body
+providing additional information about the error should be returned. For new endpoints, the response body should look
+like this:
+
+``` JavaScript
+{
+    "message": "error details here"
```

Comment:
Since I had to review this bit of UI code recently, are any of [these](https://github.com/cyverse-de/ui/blob/master/de-lib/src/main/java/org/iplantc/de/client/models/errorHandling/SimpleServiceError.java) keys potentially also standard?  In particular, the `error_code` key which the UI was sometimes using to check for [specific codes](https://github.com/cyverse-de/ui/blob/master/de-lib/src/main/java/org/iplantc/de/client/models/errorHandling/ServiceErrorCode.java), and also in [sonora](https://github.com/cyverse-de/sonora/blob/master/src/components/utils/error/errorCode.js) a bit as well.

To be clear, not saying I think any of those keys should actually be standard. Just curious where the backend folks stand on that.

Reply:
That's a good question. Several of our services use the `error_code` field to provide a machine readable error identifier, which can be really useful in cases where the UI can identify and respond to some errors. I don't know that we'd want to _require_ that field, but it might be useful to have it mentioned as an optional field in the guidelines.

I don't know of any places where the `action` and `status` fields are being used, but I'd be open to mentioning them if they're still useful.

Reply:
I updated this section to also mention the `error_code` field. I'll omit the `action` and `status` fields for now.